mono/metadata/class.c

```diff
@@ -760,18 +829,85 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
 	}
 	case MONO_TYPE_PTR: {
 		MonoType *nt, *inflated = inflate_generic_type (image, type->data.type, context, error);
-		if (!inflated || !mono_error_ok (error))
+		if ((!inflated && !changed) || !mono_error_ok (error))
 			return NULL;
+		if (!inflated && changed)
+			return type;
 		nt = mono_metadata_type_dup (image, type);
 		nt->data.type = inflated;
 		return nt;
 	}
 	default:
-		return NULL;
+		if (!changed)
+			return NULL;
+		else
+			return type;
 	}
 	return NULL;
 }
 
+static MonoType*
+inflate_generic_custom_modifiers (MonoImage *image, const MonoType *type, MonoGenericContext *context, MonoError *error)
+{
+	MonoType *result = NULL;
+	g_assert (type->has_cmods);
+	int count = mono_type_custom_modifier_count (type);
+	gboolean changed = FALSE;
+
+	size_t aggregate_size = mono_sizeof_aggregate_modifiers (count);
+	MonoAggregateModContainer *candidate_mods = g_alloca (aggregate_size);
+	memset (candidate_mods, 0, aggregate_size);
+	candidate_mods->count = count;
+
+	for (int i = 0; i < count; ++i) {
+		gboolean required;
+		MonoType *cmod_old = mono_type_get_custom_modifier (type, i, &required, error);
+		goto_if_nok (error, leave);
+		MonoType *cmod_new = inflate_generic_type (NULL, cmod_old, context, error);
+		goto_if_nok (error, leave);
+		if (cmod_new)
+			changed = TRUE;
+		candidate_mods->modifiers [i].required = required;
+		candidate_mods->modifiers [i].type = cmod_new;
+	}
+
+	if (changed) {
+		/* if we're going to make a new type, fill in any modifiers that weren't affected by inflation with copies of the original values. */
+		for (int i = 0; i < count; ++i) {
+			if (candidate_mods->modifiers [i].type == NULL) {
+				candidate_mods->modifiers [i].type = mono_metadata_type_dup (NULL, mono_type_get_custom_modifier (type, i, NULL, error));
+					
+				/* it didn't error in the first loop, so should be ok now, too */
+				mono_error_assert_ok (error);
+			}
+		}
+	}
+#ifdef DEBUG_INFLATE_CMODS
+	if (changed) {
+		char *full_name = mono_type_full_name ((MonoType*)type);
+		printf ("\n\n\tcustom modifier on '%s' affected by subsititution\n\n\n", full_name);
+		g_free (full_name);
+	}
+#endif
+	if (changed) {
+		MonoType *new_type = g_alloca (mono_sizeof_type_with_mods (count, TRUE));
```

Comment:
Can you add some assertions around this type of allocation (around the maximum return value of "mono_sizeof_type_with_mods") so that if this explodes the alt stack or something, it won't be a big debugging session?